[EIP-3742] - Standard for multi party contract (#3742)
ethereum/EIPs: EIPS/eip-multi_party_contract.md

---
eip: 3742
title: Multi Party Contract Standard
description: A standard interface for approving/executing contract transactions in a democratic way and for co-owning assets like ERC20, ERC721, ETH etc. 
author: Saurabh Santhosh (@saurabhsanthosh)
discussions-to: https://ethereum-magicians.org/t/contract-standard-for-co-owning-assets/6877
type: Standards Track
category: ERC
status: Draft
created: 2021-08-15
---

## Abstract

The following standard allows for the implementation of a standard API for co-owning assets within smart contracts. This standard provides basic functionality to add members, remove members and create/execute contract transactions approved by all members of the contract. Some of the possible actions are token transferring, operator approvals and member addition/removals. It's important to note that every group of users will be a new instance of the contract having a unique address.

> 

Reviewer: ```suggestion
Please remove the extraneous spacing in this document.
```

OPTIONAL - This method can be used if the developer wants his/her contracts to have this functionality,
but interfaces and other contracts MUST NOT expect these values to be present.

``` js
function approveERC721(address _erc721Contract, address _spender, uint256 _tokenId) external onlySelf returns (bool success)
```

### setApprovalForAllERC721

Approve or remove `operator` as an operator for the caller. This method should be external with `onlySelf` modifier and will be executed only if it's approved by the group. (Approval is defined by the `isActionApproved` method)

``` js
function setApprovalForAllERC721(address operator, bool _approved) external onlySelf returns (bool success)
```

We can have any method here we want to execute based on approval. We just have to make sure they are internal and can only be executed
via `executeAction` function

### Events

#### ActionStateChanged

MUST trigger when actions are created, approved, executed.

``` js
event ActionStateChanged(uint256 _actionId, address _from, uint256 _state)
```

> ## Implementation

Reviewer: ```suggestion
## Reference Implementation
```

> ---

Reviewer: You will need to add a `discussions-to` link for your EIP to be merged as a draft.

Author: Added the discussions-to

> eip: 

Reviewer 2: ```suggestion
eip: 3742
```

title: Multi Party Contract Standard

> author: Saurabh Santhosh <saurabhsanthosh@gmail.com>

Reviewer 2: ```suggestion
author: Saurabh Santhosh (@saurabhsanthosh), Saurabh Santhosh <saurabhsanthosh@gmail.com>
```
or
```suggestion
author: Saurabh Santhosh (@saurabhsanthosh)
```

At least one author should be a GitHub handle.

type: Standards Track
category: ERC
status: Draft
created: 2021-08-15
---

> ## Simple Summary

Reviewer: We recently deprecated the "Simple Summary" in favor of a description field in the preamble. Please see the [EIP template ](https://github.com/ethereum/EIPs/blob/master/eip-template.md)for an example.

A standard interface for co-owning assets like ERC20, ERC721, ETH etc.
and to approve/execute contract transactions in a democratic way

## Abstract

> The following standard allows for the implementation of a standard API for co-owning assets within smart contracts. This standard provides basic functionality to addMembers, removeMembers and create/execute contract transactions approved by all members of the contract. Some of the possible actions are token transferring, operator approvals and member addition/removals.Its important to note that every Group of users will be a new instance of the contract having a unique address.

Reviewer: ```suggestion
The following standard allows for the implementation of a standard API for co-owning assets within smart contracts. This standard provides basic functionality to add members, remove members and create/execute contract transactions approved by all members of the contract. Some of the possible actions are token transferring, operator approvals and member addition/removals. It's important to note that every group of users will be a new instance of the contract having a unique address.
```

```

We can have any method here we want to execute based on approval. We just have to make sure they are internal and can only be executed
via `executeAction` function

### Events

#### ActionStateChanged

MUST trigger when actions are created, approved, executed.

``` js
event ActionStateChanged(uint256 _actionId, address _from, uint256 _state)
```

## Implementation

There may already be existing contracts doing similar actions deployed on the Ethereum network.
Hopefully with single standard, it will make it easier for contract developers to handle the above mentioned
scenarios in efficient and secure manner

#### Example implementations are available at

- [README](https://github.com/saurabhsanthosh/ethereum_EIPs/blob/master/contracts/erc27/eip-27.md)
- [Interface](https://github.com/saurabhsanthosh/ethereum_EIPs/blob/master/contracts/erc27/IERC27.sol)
- [Implementation](https://github.com/saurabhsanthosh/ethereum_EIPs/blob/master/contracts/erc27/ERC27.sol)
- [Extensions](https://github.com/saurabhsanthosh/ethereum_EIPs/tree/master/contracts/erc27)
- [TestCases](https://github.com/saurabhsanthosh/ethereum_EIPs/blob/master/test/erc27.js)

> ## History

Reviewer: Please remove this section, it isn't a valid EIP-1 section
